<a href="https://colab.research.google.com/github/Digitall-Ivan/Data_Analytics_course/blob/main/Ivan_Zakorchevnyi_HW_14_3_Tests_for_small_samples.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Тести для малих вибірок та пропорцій



**Завдання 1**. E-commerce компанія після редизайну сайту підозрює, що **середній час до покупки (time-to-purchase)** користувачів **збільшився**.
Історично середній час від першого заходу на сайт до покупки становив $\mu_0 = 123.7$ хвилин.

Команда аналітиків випадково обрала дані **7 покупців після редизайну**:
`128, 135, 121, 142, 126, 151, 123`

З допомогою наявних даних зʼясуйте, чи збільшився середній час покупки після редизайну.

Для цього виконайте наступні 6 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.
2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.
3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.
4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.
5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?
6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.


In [1]:
import numpy as np
from scipy import stats

# Дані вибірки (хвилини)
x = np.array([128, 135, 121, 142, 126, 151, 123], dtype=float)

mu0 = 123.7           # історичне середнє
alpha1 = 0.10         # рівень значущості 1
alpha2 = 0.05         # альтернативний рівень

# Вибіркові статистики
xbar = x.mean()
s = x.std(ddof=1)
n = x.size
df = n - 1
SE = s / np.sqrt(n)

# t-статистика (правосторонній тест, Ha: μ > μ0)
t_stat = (xbar - mu0) / SE

# Критичні значення для правостороннього тесту
tcrit_a1 = stats.t.ppf(1 - alpha1, df)
tcrit_a2 = stats.t.ppf(1 - alpha2, df)

# p-value (right-tail)
p_value = 1 - stats.t.cdf(t_stat, df)

# Повернемо результати
{
    "xbar": round(float(xbar), 4),
    "s": round(float(s), 4),
    "n": int(n),
    "SE": round(float(SE), 4),
    "df": int(df),
    "t_stat": round(float(t_stat), 4),
    "tcrit_alpha_0_10": round(float(tcrit_a1), 4),
    "tcrit_alpha_0_05": round(float(tcrit_a2), 4),
    "p_value": round(float(p_value), 4)
}


{'xbar': 132.2857,
 's': 10.9805,
 'n': 7,
 'SE': 4.1502,
 'df': 6,
 't_stat': 2.0687,
 'tcrit_alpha_0_10': 1.4398,
 'tcrit_alpha_0_05': 1.9432,
 'p_value': 0.042}

Крок 1 Гіпотези та тип тесту

H₀: μ = 123.7 (час не збільшився)

H₁: μ > 123.7 (час збільшився після редизайну)

Тест: правосторонній t-test (n менше 30, σ невідома).

Крок 2 Вибіркові статистики

*   𝑥(сер) = 132.2857
*   s = 10.9805
*   𝑛 = 7
*   𝑆𝐸 = 4.1502
*   𝑑𝑓 = 6

Крок 3 Розрахунок тесту


*   𝑡 = (𝑥(сер)-𝜇0)/𝑆𝐸 = 2.0687
*   Критичні значення (правий хвіст):
при α = 0.10:
𝑡(крит) = 1.4398,
при α = 0.05:
𝑡(крит) = 1.9432,
*   p-value (right-tail) = 0.0420

Крок 4 Рішення за p-value (α = 0.10) : p=0.042<0.10 → відхиляємо H₀

Крок 5 Висновок : Є статистично значущі докази, що середній час до покупки збільшився після редизайну.

Крок 6 Якщо α = 0.05, то p=0.042<0.05 → рішення не змінюється: також відхиляємо H₀ і робимо той самий висновок про збільшення часу

**Завдання 2.**

До спеціальної рекламної кампанії **23%** дорослих упізнавали логотип компанії. Після завершення кампанії відділ маркетингу провів опитування: з **1200** випадково відібраних дорослих **311** упізнали логотип.

Перевірте на рівні значущості **3%** ($\alpha=0.03$), чи дають ці дані достатні підстави стверджувати, що **тепер більше ніж 23%** дорослих упізнають логотип компанії. Для розвʼязку використовуйте бібліотеку `statsmodels`.

Зробіть висновок, чи зросла впізнаваність логотипу.

Додатково, обчисліть довірчий інтревал на заданому рівні значущості і проінтерпретуйте текстом - як він додатково пояснює прийняте нами рішення?

In [2]:
import numpy as np
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

# Вхідні дані
p0 = 0.23           # історична пропорція
successes = 311
n = 1200
alpha = 0.03

# z-тест для пропорцій (Ha: p > p0)
z_stat, p_value = proportions_ztest(successes, n, value=p0, alternative="larger")

# Довірчий інтервал для пропорції (1 - alpha)
confint_low, confint_high = proportion_confint(successes, n, alpha=alpha, method="normal")

{
    "sample_proportion": round(successes/n, 4),
    "z_stat": round(float(z_stat), 4),
    "p_value": round(float(p_value), 4),
    "confint_low": round(float(confint_low), 4),
    "confint_high": round(float(confint_high), 4)
}


{'sample_proportion': 0.2592,
 'z_stat': 2.3058,
 'p_value': 0.0106,
 'confint_low': 0.2317,
 'confint_high': 0.2866}

Крок 1 Вихідні дані задачі


*   Початкова частка (до кампанії): 𝑝0 = 0.23
*   Вибіркова частка після кампанії: 𝑝^ = 311/1200 = 0.2592
*   Розмір вибірки: 𝑛 = 1200
*   Рівень значущості: α= 0.03

Крок 2 Розрахунки

*   𝑧 = 2.3058
*   Р-value=0.0106
*   Довірчий інтервал (97%): [0.2317; 0.2866]

Крок 3 Висновок : Оскільки 𝑝-value = 0.0106 < 𝛼 = 0.03, ми відхиляємо H₀ та приймаємо альтернативну гіпотезу, тобто, після кампанії частка людей, які впізнають логотип, статистично значуще більша за 23%.

Крок 4 - Інтерпретація довірчого інтервалу : 97%-й довірчий інтервал для частки становить від 23.17% до 28.66%. Це означає, що з ймовірністю 97% істинне значення нової частки впізнаваності знаходиться у цьому діапазоні.Оскільки вся нижня межа (0.2317) трохи перевищує 𝑝0 = 0.23, це додатково підтверджує зростання впізнаваності.